management-node/server.go

```diff
@@ -0,0 +1,530 @@
+		for _, task := range req.Tasks {
+			task.State = common.TaskStateDead
+			task.Id = uuid.New().String()
+		}
+
+		if err := md.setTasksToDb(req.Tasks); err != nil {
+			log.Error(err.Error())
+			return nil, err
+		}
+
+		return req, nil
+	}
+
+	// msg type for internal use
+	type msg struct {
+		err  error
+		task *pb.Task
+	}
+
+	// channel for connecting onRsp handlers
+	// to the current go routine
+	messanger := make(chan msg, 1)
+
+	for i, task := range req.Tasks {
+
+		// all the ids are generated by management node
+		id := uuid.New().String()
+		workerId := workers[0]
+
+		// RR worker id calculation
+		if Config.SchedulerAlgo == SchedAlgoRR {
+			workerId = workers[int(math.
+				Mod(float64(i),
+					float64(len(workers))))]
+		} else if Config.SchedulerAlgo == SchedAlgoRand {
+			s := rand.NewSource(time.Now().UnixNano())
+			r := rand.New(s)
+
+			workerId = workers[r.Intn(len(workers))]
+
+		}
+
+		// set task state to pending
+		task.Id = id
+		task.State = common.TaskStatePending
+		task.WorkerId = workerId
+
+		if err := md.SetToDb(task, EtcdTaskPrefix+id); err != nil {
+			log.Error(err.Error())
+			return nil, err
+		}
+
+		// setup handlers
+		// copy object to handler
+		t := *task
+		onRspHandler := func(rsp *pb.WorkerRsp) {
+
+			// TBD: validate reply
+
+			t.State = rsp.Reply
+			var err error
+			if err = md.SetToDb(&t, EtcdTaskPrefix+t.Id); err != nil {
+				log.Error(err.Error())
+			}
+
+			messanger <- msg{
+				task: &t,
+				err:  err,
+			}
+		}
+
+		onTimerExpiredHandler := func() {
+			err := fmt.Errorf("timer is expired for rsp on task %s", task.Id)
+			log.Error(err.Error())
+			messanger <- msg{
+				err:  err,
+				task: &t,
+			}
+		}
+
+		// combine parameters with cmd definition
+		params := append([]string{task.Cmd}, task.Parameters...)
+
+		// prepare request
+		req := pb.MgmtReq{
+			Id:     task.Id,
+			Method: common.WorkerNodeRPCExec,
+			Params: params,
+		}
+
+		// task is scheduled
+		// setup dead timeout
+		md.SetTaskAndRunDeadTimeout(&Task{
+			task: task,
+			stop: make(chan struct{}, 1),
+		})
+
+		// sending to worker node
+		md.workerNodePool[workerId].
+			SendWithHandlerTimeout(req, onRspHandler,
+				onTimerExpiredHandler,
+				time.Duration(Config.SilenceTimeout))
+	}
+
+	// prepare task objects to response to schedctl
+	tasks := []*pb.Task{}
+	for _ = range req.Tasks {
+		m := <-messanger
+
+		// check if it responded with error
+		if m.err != nil {
+			log.Error(m.err)
+
+			// set error to worker error field
+			m.task.WorkerError = m.err.Error()
+			if err := md.SetToDb(m.task, EtcdTaskPrefix+m.task.Id); err != nil {
+				log.Error(err.Error())
+				return nil, err
+			}
+			tasks = append(tasks, m.task)
+			continue
+		}
+
+		tasks = append(tasks, m.task)
+	}
+
+	close(messanger)
+
+	return &pb.TaskList{Tasks: tasks}, nil
+
+}
+
+// SetToDb stores subj to etcd by id
+func (md *ManagementNode) SetToDb(subj interface{}, id string) error {
```

Comment:
I've move all this code to another package right now `ManagementNode` node mixes business logic and data.

Reply:
ok, will do it in future